### ISSUE TITLE
If error WARNING is enable fopen break in case of connection failure

### ISSUE THREAD

**Issue by bobonov** [yiisoft/yii2-httpclient#38]
A better error handling is needed. In case of connection timeout fopen does a warning which is intercepted by yii2 and break execution
fopen(http://myfakeurl.com): failed to open stream: Connection timed out

yiisoft/yii2-httpclient/StreamTransport.php line 59

$stream = fopen($url, 'rb', false, $context);

I know that in production warning should be disabled but errors should be better handled, especially if is a connection error I  need to know it so I can inform in some way the user.

Anyway during development I need warning enable and putting in the code an error_reporting(E_ERROR | E_PARSE) just before httpclient call is not nice.

**Comment by SilverFire** [comment]
Yeah, it can be enhanced. Would you like to work on a PR?

**Comment by klimov-paul** [comment]
What is wrong with `try catch` block using?

**Comment by bobonov** [comment]
try cath do not intercept warning but exception and fopen does a warning.
So when warning is enabled in error reporting it pass the try catch but yii intercept it and stop the execution (exactly like a warning for using an undeclared var)

If you want to reproduce it easily you can try something like this (being sure nothing respond on 8080):
```php
        $client = new Client(['baseUrl' => 'http://127.0.0.1:8080']);
        $response = $client->createRequest()->setMethod('get')->setUrl('fakeUrl')->send();
```

To make the httpclient code working as exception I think we need something like this:
```php
// this will trow an exception whatever is the problem
set_error_handler(function($errno, $errstr, $errfile, $errline, array $errcontext) {
    throw new ErrorException($errstr, 0, $errno, $errfile, $errline);
});
// we save the current error level
$error=ini_set('error_reporting');
// we add the warning
error_reporting($error | E_WARNING);
$stream = fopen($url, 'rb', false, $context);
// We restore the error level
error_reporting($error);
// We restore the error handler
restore_error_handler();
```
This will trow an exception, so we can use httpclinet in try catch.
But this is a stubb. httpclient need an internal error handler which map error code from different driver to the same code.
curl and fopen give different errors for the same problem but httpclient should give always the same message and code.

**Comment by klimov-paul** [comment]
> try cath do not intercept warning but exception and fopen does a warning.

Yii built-in error handler already converts PHP errors and warnings into exceptions.

**Comment by klimov-paul** [comment]
https://github.com/yiisoft/yii2/blob/master/framework/base/ErrorHandler.php#L166

**Comment by bobonov** [comment]
Wheel I tried to use try catch (this was how I wrote the code first...) but it didn't catch anything
```php
        try {
            $client = new Client(['baseUrl' => 'http://localhost:8080']);
            $response = $client->createRequest()
                ->setMethod('get')->setUrl('fakeService')->send();
        } catch (Exception $e) {
            die('my exception handling: '.  $e->getMessage(). "\n");
        }
```
So either yii or httpclient fail somewhere in hadling fopen warning.
Anyway is still true that httpclient need an internal error handler that give same error message whatever you use fopen or curl

**Comment by klimov-paul** [comment]
Does catch block is written correct?
You ensure you are catching base exception class `\Exception`:

```php
} catch (\Exception $e) { //use gloabl namespace
```

**Comment by bobonov** [comment]
Yes my bad forgot to put \ before Exception, thank you to notice out. When you keep reading your code many time you start to miss elementary problems.

Anyway the enhancement is still valid.

**Comment by klimov-paul** [comment]
> httpclient need an internal error handler which map error code from different driver to the same code.

I am unsure it is possible. Do you have a particular suggestion for this?

**Comment by bobonov** [comment]
Didn't analyzed the situation. I'll try to look at how curl and fopen handle the different situations and eventually propose something

**Comment by klimov-paul** [comment]
Resolved by commit 3e90b71243b9025c8d6360dd5b755371e23f0e28
